misc/codepoint_width.c

```diff
@@ -0,0 +1,758 @@
+    0xc0, 0x40, 0x40,
+};
+static const uint32_t s_joinRules[2][16] = {
+    {
+        0b00000011110011111111111111001111,
+        0b00001111111111111111111111111111,
+        0b00000011110011111111111111001111,
+        0b00000011110011111111111101001111,
+        0b00000000000000000000000000001100,
+        0b00000011110000001100001111001111,
+        0b00000011110011110000111111001111,
+        0b00000011110011110011111111001111,
+        0b00000011110011110000111111001111,
+        0b00000011110011110011111111001111,
+        0b00000011000011111111111111001111,
+        0b00000011110011111111111111001111,
+        0b00000011110011111111111111001111,
+        0b00000000110011111111111111001111,
+        0b00000000000000000000000000000000,
+        0b00000000000000000000000000000000,
+    },
+    {
+        0b00000011110011111111111111001111,
+        0b00001111111111111111111111111111,
+        0b00000011110011111111111111001111,
+        0b00000011110011111111111111001111,
+        0b00000000000000000000000000001100,
+        0b00000011110000001100001111001111,
+        0b00000011110011110000111111001111,
+        0b00000011110011110011111111001111,
+        0b00000011110011110000111111001111,
+        0b00000011110011110011111111001111,
+        0b00000011000011111111111111001111,
+        0b00000011110011111111111111001111,
+        0b00000011110011111111111111001111,
+        0b00000000110011111111111111001111,
+        0b00000000000000000000000000000000,
+        0b00000000000000000000000000000000,
+    },
+};
+static int ucdLookup(const char32_t cp)
+{
+    const uint16_t s0 = s_stage0[cp >> 11];
+    const uint16_t s1 = s_stage1[s0 + ((cp >> 6) & 31)];
+    const uint16_t s2 = s_stage2[s1 + ((cp >> 3) & 7)];
+    const uint8_t s3 = s_stage3[s2 + ((cp >> 0) & 7)];
+    return s3;
+}
+static int ucdGraphemeJoins(const int state, const int lead, const int trail)
+{
+    const int l = lead & 15;
+    const int t = trail & 15;
+    return (s_joinRules[state][l] >> (t * 2)) & 3;
+}
+static bool ucdGraphemeDone(const int state)
+{
+    return state == 3;
+}
+static int ucdToCharacterWidth(const int val)
+{
+    return val >> 6;
+}
+// clang-format on
+// End of generated code
+
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+int term_disp_width(bstr str)
+{
+    static const int ambiguous_width = 1;
```

Comment:
This is expected. Do you have examples where this matters? If you read next line, 1 is more accurate fallback for ambiguous width. We use 1 width by default, but once we see `0xFE0F` to convert those unqualified Emojis we promote them to 2 width. That said I think it works as expected and best we can do. If you have cases where it breaks, please share I will adjust accordingly.

Reply:
Are we still using those circles? Didn't we want to remove them?

Hmm, one small thing is that msg.c will actually work better when we underestimate the width, instead of overestimating. Basically in our case current way will work better in practice, equally bad, but "better" in the sense it will not eat lines and instead duplicate.

Reply:
> I didn't have a strong opinion on the usage of circles instead of ASCII symbols, as long as the same width is obtained for active and inactive states.

No, guido explained me the situation on IRC. I recalled topic wrong.

> If overestimating causes lines being eaten which can't be easily solved, then this is OK for now, but I would consider making it an option later.

I will keep it as is if you don't mind, this PR is very hot already, I don't want to add any more changes here. We can improve later, if we manage to merge this guy first. Hope you understand.